doc/source/dev/testing.rst

```diff
@@ -0,0 +1,106 @@
+##############
+Testing in ODL
+##############
+
+ODL is tested using pytest_ and has four main types of tests and verification measures that are supposed to test the code in different ways. These are listed below along with the command to run them.
+
+============  =====================  =======
+Name          Command                Description
+============  =====================  =======
+Unittests     `pytest`               Test "micro-features" of the code, like testing that each parameter combination works, that the correct exceptions are raced and that the code works correctly in corner cases and with a wide range of input.
+Largescale    `pytest --largescale`  Verify that the functionality works well in more complicated and realistic conditions.
+Doctests      `pytest`               Meant as simple examples of how to use a functionality for users reading the documentation.
+Examples      `pytest --examples`    Copy-paste friendly examples on how to use a function in a more complete context.
+============  =====================  =======
+
+Unittests
+~~~~~~~~~
+All unit-tests in ODL are contained in the `odl/tests` folder, where each ODL sub-package has a test file of its own. Any major ODL functionality should have unittests covering all of the use cases that are implied in the documentation. In addition to this, the tests should be quick to run, preferably at most a few milliseconds per tests. This is because the tests should be run as often as possible in order to verify that no functionality has been broken when adding new code or refactoring old code.
+
+A short example of testing a function is given below, for more information consult the pytest documentation and look at existing tests in the test folder.
+
+.. code:: python
+
+    import numpy as np
+    import pytest
+
+
+    def myfunction(x):
+        """Convert ``x`` to a integer and add 1"""
+        return int(x) + 1
```

Comment:
This is nice and simple, but IMO it tests the wrong thing and gives a slightly wrong impression of what testing is supposed to achieve. The problem is that the tests are entirely testing the builtin `int` constructor, which is not exactly the purpose of testing. Testing against bad input makes more sense if our own implementation contains checks and we want to make sure they are effective. Of course it's also valid to check that the functionality we rely upon behaves in the way we expect, but it would be nice to have at least one self-implemented aspect that is checked.